### PR TITLE
Dont throw Invalid DAG error for valid pipeline

### PR DESCRIPTION
**Pull request by CuriousVini**
Issue: https://issues.cask.co/browse/HYDRATOR-905
Build: http://builds.cask.co/browse/CDAP-DUT4779-4

### CODE REVIEW

**Comment by anew** [comment]
Please explain what caused the bug and what the approach to fix it?
Also, please add minimal comments in the code explaining what it is doing. 
Also, please add tests for all the corner cases covered by this fix.

**Comment by CuriousVini** [comment]
@anew Addressed comments. Added description about the bug in [jira](https://issues.cask.co/browse/HYDRATOR-905)

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java [hunk 1]
```diff
@@ -239,12 +241,60 @@ private void isolate(String node, Set<String> addedAlready) {
     }
 
     Set<String> remainingSources = Sets.intersection(remainingNodes, possibleNewSources);
+    Set<String> processedNodes = new HashSet<>();
+    // Since there can be remaining sources from subdags which don't overlap, they should be split as seperate subdags.
+    // For that, we first create all subdags from remaining sources and keep track of accessible nodes for those
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java]: it would help to draw a diagram in the comments to illustrate what is happening. Otherwise this very hard to describe. You could re-use the one from the test case.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java [hunk 2]
```diff
@@ -239,12 +241,60 @@ private void isolate(String node, Set<String> addedAlready) {
...
+    // subdags. Then we merge overlapping subdags
     if (!remainingSources.isEmpty()) {
-      dags.add(subsetFrom(remainingSources, possibleNewSinks));
+      // source -> [ nodes accessible by the source ]
+      Map<String, Set<String>> nodesAccessibleBySources = new HashMap<>();
+
+      for (String remainingSource : remainingSources) {
+        Dag remainingNodesDag = subsetFrom(ImmutableSet.of(remainingSource), possibleNewSinks);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java]: there is already a convenience method `subsetFrom(remainingSource, possibleNewSinks);`

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java [hunk 3]
```diff
@@ -239,12 +241,60 @@ private void isolate(String node, Set<String> addedAlready) {
...
+        nodesAccessibleBySources.put(remainingSource, accessibleFrom(remainingSource, remainingNodesDag.getSinks()));
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java]: you can just use remainingNodesDag.getNodes()

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java [hunk 4]
```diff
@@ -239,12 +241,60 @@ private void isolate(String node, Set<String> addedAlready) {
...
+      }
+
+      for (String remainingSource : remainingSources) {
+        if (!processedNodes.contains(remainingSource)) {
+          Set<String> remainingAccessibleNodes = nodesAccessibleBySources.get(remainingSource);
+          // Island of all the remaining nodes
+          Set<String> islandNodes = new HashSet<>();
+          islandNodes.addAll(remainingAccessibleNodes);
+          for (String otherSource : remainingSources) {
+            if (!remainingSource.equalsIgnoreCase(otherSource)) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java]: shouldn't use ignoreCase. We check that there are no duplicate nodes, but don't think we do it in a case-insensitive way.

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java]: nit: a little easier to read if you do

```
if (remainingSource.equals(otherSource)) {
  continue;
}
```

since it removes one layer of nesting

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java [hunk 5]
```diff
@@ -239,12 +241,60 @@ private void isolate(String node, Set<String> addedAlready) {
...
+      // source -> [ nodes accessible by the source ]
+      Map<String, Set<String>> nodesAccessibleBySources = new HashMap<>();
+
+      for (String remainingSource : remainingSources) {
+        Dag remainingNodesDag = subsetFrom(ImmutableSet.of(remainingSource), possibleNewSinks);
+        nodesAccessibleBySources.put(remainingSource, accessibleFrom(remainingSource, remainingNodesDag.getSinks()));
+      }
+
+      for (String remainingSource : remainingSources) {
+        if (!processedNodes.contains(remainingSource)) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java]: nit: a little easier to read if you do

```
if (processedNodes.contains(remainingSource)) {
  continue;
}
```

since it removes one layer of nesting

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java [hunk 6]
```diff
@@ -239,12 +241,60 @@ private void isolate(String node, Set<String> addedAlready) {
...
     }
     return dags;
   }
 
+  /**
+   * Creates dag from provided list of island nodes
+   * @param islandNodes list of island nodes
+   * @return Dag with connections among island nodes
+   */
+  private Dag createDag(Set<String> islandNodes) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java]: It would be better to move this into the Dag class as a general `createSubDag(Set<String> nodes)` method, since it is creating another Dag from a subset of the nodes in the current Dag. 'island' doesn't really describe anything.

**Comment by albertshau** [comment]
logic looks good, but have a few minor comments.

**Comment by CuriousVini** [comment]
@anew @albertshau Addressed comments. Please take another when you get a chance, thanks!!

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java [hunk 7]
```diff
@@ -239,8 +241,60 @@ private void isolate(String node, Set<String> addedAlready) {
     }
 
     Set<String> remainingSources = Sets.intersection(remainingNodes, possibleNewSources);
+    Set<String> processedNodes = new HashSet<>();
+    // Since there can be remaining sources from subdags which don't overlap, they should be split as seperate subdags.
+    // For example:
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java]: nit: these comments can all be part of the larger comment `/* */`.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java [hunk 8]
```diff
@@ -239,8 +241,60 @@ private void isolate(String node, Set<String> addedAlready) {
...
+                  |--- n3(r) ---|
+             n2 --|             |-- n5(r) --|
+                    n4 ---------|           |
+                                            |-- n7(r) -- n8
+                                            |
+                                n6 ---------|
+                                            |
+                                n9 ---------|
+
+      Here, n6, n9, n7.connector should be in the same subdag and n4, n5.connector should be another subdag.
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java]: Should mention that at this point, remainingSources are n4, n6, and n9. Would also be better to put the connector nodes in the diagram above.

**Comment by albertshau** [comment]
minor comments about clarifying comments. LGTM, fix then merge.

**Comment by CuriousVini** [comment]
Addressed minor comments. Build is green; squashing and merging.
